src/client/ds3/object.c

```diff
@@ -0,0 +1,374 @@
+{
+	int         rc = 0;
+	char       *path;
+	dfs_obj_t  *parent = NULL;
+	char       *file_start;
+	const char *file_name;
+	const char *parent_path = NULL;
+	char       *lookup_path = NULL;
+
+	D_STRNDUP(path, key, DS3_MAX_KEY - 1);
+	if (path == NULL) {
+		return -ENOMEM;
+	}
+
+	file_start = strrchr(path, '/');
+	file_name  = path;
+	if (file_start != NULL) {
+		*file_start = '\0';
+		file_name   = file_start + 1;
+		parent_path = path;
+	}
+
+	if (parent_path != NULL) {
+		D_ALLOC_ARRAY(lookup_path, DS3_MAX_KEY);
+		if (lookup_path == NULL) {
+			rc = ENOMEM;
+			goto err_path;
+		}
+
+		strcpy(lookup_path, "/");
+		strcat(lookup_path, path);
+		rc = dfs_lookup(ds3b->dfs, lookup_path, O_RDWR, &parent, NULL, NULL);
+		if (rc != 0) {
+			goto err_parent;
+		}
+	}
+
+	rc = dfs_remove(ds3b->dfs, parent, file_name, false, NULL);
+
+err_parent:
+	if (parent)
+		dfs_release(parent);
+	if (lookup_path)
+		D_FREE(lookup_path);
+err_path:
+	D_FREE(path);
+	return -rc;
+}
+
+int
+ds3_obj_write(void *buf, daos_off_t off, daos_size_t *size, ds3_bucket_t *ds3b, ds3_obj_t *ds3o,
+	      daos_event_t *ev)
+{
+	d_sg_list_t wsgl;
+	d_iov_t     iov;
+
+	if (ds3b == NULL || buf == NULL || ds3o == NULL)
+		return -EINVAL;
+
+	d_iov_set(&iov, buf, *size);
+	wsgl.sg_nr   = 1;
+	wsgl.sg_iovs = &iov;
+	return -dfs_write(ds3b->dfs, ds3o->dfs_obj, &wsgl, off, ev);
+}
+
+int
+ds3_obj_mark_latest(const char *key, ds3_bucket_t *ds3b)
+{
+	int         rc = 0;
+	char       *path;
+	dfs_obj_t  *parent = NULL;
+	char       *file_start;
+	const char *file_name;
+	const char *parent_path = NULL;
+	char       *link_name   = NULL;
+	int         name_length;
+	char       *suffix_start;
+	dfs_obj_t  *link;
+
+	if (ds3b == NULL || key == NULL)
+		return -EINVAL;
+
+	if (ends_with(key, LATEST_INSTANCE_SUFFIX)) {
+		D_ERROR("Creating an object that ends with %s is not allowed.\n",
+			LATEST_INSTANCE_SUFFIX);
+		return -EINVAL;
+	}
+
+	D_STRNDUP(path, key, DS3_MAX_KEY - 1);
+	if (path == NULL) {
+		return -ENOMEM;
+	}
+
+	file_start = strrchr(path, '/');
+	file_name  = path;
+	if (file_start != NULL) {
+		*file_start = '\0';
+		file_name   = file_start + 1;
+		parent_path = path;
+	}
+
+	char *lookup_path = NULL;
+	if (parent_path != NULL) {
+		D_ALLOC_ARRAY(lookup_path, DS3_MAX_KEY);
+		if (lookup_path == NULL) {
+			rc = ENOMEM;
+			goto err_path;
+		}
+
+		strcpy(lookup_path, "/");
+		strcat(lookup_path, path);
+		rc = dfs_lookup(ds3b->dfs, lookup_path, O_RDWR, &parent, NULL, NULL);
+		if (rc != 0) {
+			goto err_parent;
+		}
+	}
+
+	// Build link name
+	D_ALLOC_ARRAY(link_name, DS3_MAX_KEY);
+	if (link_name == NULL) {
+		rc = ENOMEM;
+		goto err_parent;
+	}
+
+	// Copy name without instance
+	name_length  = strlen(file_name);
+	suffix_start = strrchr(file_name, '[');
+	if (suffix_start != NULL) {
+		name_length = suffix_start - file_name;
+	}
+
+	strncpy(link_name, file_name, name_length);
+	strcat(link_name, LATEST_INSTANCE_SUFFIX);
```

Comment:
Check that `file_name` plus `LATEST_INSTANCE_SUFFIX` won't overrun?

Reply:
That's a good point. I changed DS3_MAX_KEY to be DS3_MAX_KEY_BUFF - 8 to account for this, and added length checks.